segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/TableIterator.java

```diff
@@ -0,0 +1,326 @@
+    private final Executor executor;
+    private final AtomicReference<Iterator<TableBucket>> currentBatch = new AtomicReference<>();
+    private final AtomicBoolean inProgress = new AtomicBoolean(false);
+
+    //endregion
+
+    //region AsyncIterator Implementation
+
+    @Override
+    public CompletableFuture<T> getNext() {
+        // Verify no other call to getNext() is currently executing.
+        Preconditions.checkState(this.inProgress.compareAndSet(false, true), "Another call to getNext() is in progress.");
+        return getNextBucket()
+                .thenCompose(bucket -> {
+                    this.inProgress.set(false);
+                    if (bucket == null) {
+                        // We are done.
+                        return CompletableFuture.completedFuture(null);
+                    } else {
+                        // Convert the TableBucket into the desired result.
+                        return this.resultConverter.apply(bucket);
+                    }
+                })
+                .exceptionally(ex -> {
+                    this.inProgress.set(false);
+                    throw new CompletionException(ex);
+                });
+    }
+
+    /**
+     * Gets the next {@link TableBucket} in the iteration. This will either be served directly from the cached batch
+     * of {@link TableBucket}s or a new invocation to the underlying indexHashIterator will be performed to fetch the next
+     * {@link TableBucket}.
+     */
+    private CompletableFuture<TableBucket> getNextBucket() {
+        val fromBatch = getNextBucketFromExistingBatch();
+        if (fromBatch != null) {
+            return CompletableFuture.completedFuture(fromBatch);
+        }
+
+        val canContinue = new AtomicBoolean(true);
+        return Futures
+                .loop(
+                        canContinue::get,
+                        this::fetchNextTableBuckets,
+                        canContinue::set,
+                        this.executor)
+                .thenApply(v -> getNextBucketFromExistingBatch());
+    }
+
+    /**
+     * Gets the next {@link TableBucket} in the iteration from the cached batch. If the batch is exhausted, returns null.
+     */
+    private TableBucket getNextBucketFromExistingBatch() {
+        val batch = this.currentBatch.get();
+        if (batch != null) {
+            TableBucket next = batch.next();
+            if (!batch.hasNext()) {
+                this.currentBatch.set(null);
+            }
+
+            return next;
+        }
+
+        return null;
+    }
+
+    /**
+     * Fetches the next set of {@link TableBucket}s from the indexHashIterator.
+     */
+    private CompletableFuture<Boolean> fetchNextTableBuckets() {
+        return this.indexHashIterator
+                .getNext()
+                .thenApplyAsync(indexHashes -> {
+                    List<TableBucket> buckets = toBuckets(indexHashes);
+                    if (buckets == null) {
```

Comment:
Why do we need both `buckets == null` and `buckets.isEmpty()`? Couldn't we represent the end of an iteration with `buckets.isEmpty()`?

Reply:
According to the `AsyncIterator` contract, `null` means end of iteration. 

`isEmpty` is simply defensive programming. If we ever encounter that from the upstream attribute iterator, we can just skip over it and request the next one.